mgmt: mcumgr: Fix Bluetooth transport issues

nordicjm opened the pull request:
This fixes issues with the Bluetooth SMP transport whereby deadlocks could arise from connection references being held in long-lasting mcumgr command processing functions.

Fixes #51653
Fixes #51714

nordicjm committed:
mgmt: mcumgr: Fix Bluetooth transport issues

This fixes issues with the Bluetooth SMP transport whereby deadlocks
could arise from connection references being held in long-lasting
mcumgr command processing functions.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented on include/zephyr/mgmt/mcumgr/smp.h:
```diff
@@ -79,6 +80,19 @@ typedef void (*smp_transport_ud_free_fn)(void *ud);
+ * @brief Function for checking if queued data is still valid.
+ *
+ * This function is used to check if queued SMP data is still valid e.g. on a remote device
+ * disconnecting, this is triggered when X is called
```
Find x (10 marks)

nordicjm commented on subsys/mgmt/mcumgr/transport/smp_bt.c:
```diff
@@ -460,7 +492,12 @@ static int smp_bt_tx_pkt(struct net_buf *nb)
 			sent = true;
 
-			k_sem_take(&smp_notify_sem, K_FOREVER);
+			if (last == false) {
```
Junk this if it is no longer needed

jori-nordic commented on subsys/mgmt/mcumgr/transport/smp_bt.c:
```diff
@@ -502,41 +551,84 @@ static void disconnected(struct bt_conn *conn, uint8_t reason)
+
+	cpd = conn_param_data_get(conn);
+
+	if (ud->conn == conn && (cpd->id == 0 || cpd->id == ud->id)) {
```
If I understand correctly, this is checking that the smp buffer belongs to the conn obj passed in `arg`, right?
Why do we need the `cpd->id == 0` part?

edit: I see this block in `connected`
```c
			if (cpd->id == 0) {
				++cpd->id;
				++next_id;
			}
```
does that mean that `id==0` has a special meaning?

nordicjm replied:
ID = 0 means unused, ID != 0 means used

de-nordic commented on subsys/mgmt/mcumgr/transport/smp_bt.c:
```diff
@@ -493,7 +514,35 @@ int smp_bt_unregister(void)
+		cpd = conn_param_data_alloc(conn);
+		k_sem_reset(&cpd->smp_notify_sem);
+
+		/* Generate an ID for this connection */
```
What are the chances that `connected` gets preempted and other `connected` instance gets called?
We can use CPU ticks here or something like that, there is very low chance that a connection will be held long enough to wrap around and get the same ticks for two different connections, and even in code executed in parallel, unless two CPUs are used, ticks should differ.
We can probably use only the lower 32bits of  `k_uptime_ticks(void)` safely.

nordicjm replied:
> I have some trouble following why this ID is needed, could you elaborate?
The ID is needed because a connection can be established, a mcumgr command can be sent, then that device can disconnect and a new device can connect - if this is a long running command it can take seconds to run, at which point without the ID check it will send the response to the new device which is invalid operation. In terms of a race condition, this is executed in the system workqueue, which runs to completion, and would need to exhaust 255 entries to loop around (which, assuming someone was connecting and disconnecting continuously, assuming it would take 30ms to do a cycle, that would take 7.65 seconds)

> What are the chances that connected gets preempted and other connected instance gets called
0 as far as I'm aware as it's called in the system workqueue which is blocked until it has completed (co-operative thread)

jori-nordic commented:
@de-nordic https://github.com/zephyrproject-rtos/zephyr/pull/51757#discussion_r1014223549

Yeah, I will bring this up with the other bluetooth maintainers, as I've seen a bunch of cases where we want to do exactly that, check that the conn object is still pointing to the connection (session?) we thought it was.
Maybe having the cpu ticks or uptime would be enough, yeah. It would also double-up as a session-length counter (until it wraps, that is). 
I'll see what I can do to have something like that as an official API sometime in the near future.

nordicjm commented on subsys/mgmt/mcumgr/transport/smp_bt.c:
```diff
@@ -502,41 +551,84 @@ static void disconnected(struct bt_conn *conn, uint8_t reason)
 	}
 
+	while (i < CONFIG_BT_MAX_CONN) {
+		k_sem_init(&conn_data[i].smp_notify_sem, 0, 1);
```
*Count of 0, the default will be 0 and maximum will be 1 (per connection). The idea being that for each connection, you can send 1 message at a time and have to wait for the sent callback semaphore to be given

jori-nordic replied:
my bad, I read too fast and didn't see the index for the semaphore

nordicjm commented on subsys/mgmt/mcumgr/smp.c:
```diff
@@ -171,6 +173,55 @@ smp_rx_req(struct smp_transport *smpt, struct net_buf *nb)
 	k_work_submit_to_queue(&smp_work_queue, &smpt->work);
 }
 
+void smp_rx_clear(struct smp_transport *zst, bool clear_all, void *arg)
```
Clear all is for future use, I've left it here so to avoid going through a stable API change. The usage would be for something like the UART transport, if a user wants to disable mcumgr then all pending queries can be cleared to use the UART for some other purpose

de-nordic commented:
> @de-nordic [#51757 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/51757#discussion_r1014223549)
> 
> Yeah, I will bring this up with the other bluetooth maintainers, as I've seen a bunch of cases where we want to do exactly that, check that the conn object is still pointing to the connection (session?) we thought it was. Maybe having the cpu ticks or uptime would be enough, yeah. It would also double-up as a session-length counter (until it wraps, that is). I'll see what I can do to have something like that as an official API sometime in the near future.

Having some unique identifier passed from BLE susbystem would be nice, because that would free us, and other subsystems, from figuring out to add additional layer of identification.
Hashes on objects does not make sense, due to statistical reasons, and random values would require recording values in use.  Value like ticks or time since system startup or first connection are increasing automatically (are monotonic, as much as type size allows before they overlap) giving unique identifier each time when picked as long as precision allows to obtain different values for closest possible connections, and as long as they have range that is able to hold value longer than time the connection information is held by a system.

de-nordic reviewed (changes requested):
Please release notes entry for MCUmgr describing addition, and to fixed bugs section.

de-nordic commented:
Question: why aren't we allocating the id in `smp_bt_chr_write` after this:
https://github.com/zephyrproject-rtos/zephyr/blob/4bae4f426ff4c5e4b84f29d8150789db1c7c569b/subsys/mgmt/mcumgr/transport/smp_bt.c#L262-L264
and before
https://github.com/zephyrproject-rtos/zephyr/blob/4bae4f426ff4c5e4b84f29d8150789db1c7c569b/subsys/mgmt/mcumgr/transport/smp_bt.c#L233

In case when reassembly is enabled this basically can be done by having static counter within function that will be incremented every time function is invoked.

Non re-assembly case is a little bit trickier because there may be multiple buffers per request, but this can probably be resolved my doing conn to id map.

nordicjm commented on subsys/mgmt/mcumgr/transport/smp_bt.c:
```diff
@@ -493,7 +513,13 @@ int smp_bt_unregister(void)
+		cpd = conn_param_data_alloc(conn);
+		cpd->id =  (uint32_t)k_uptime_ticks();
+		k_sem_reset(&cpd->smp_notify_sem);
+		__ASSERT(cpd->id != 0, "Connection generated ID must not be zero");
```
Fixed

nordicjm commented:
> Question: why aren't we allocating the id in smp_bt_chr_write after this:

The way it works here is much simpler and works for all queued packets, i.e. a device can queue up 5 commands and disconnect, it does not need different IDs for each command, they should be the same.

nordicjm commented on subsys/mgmt/mcumgr/transport/smp_bt.c:
```diff
@@ -420,7 +431,18 @@ static int smp_bt_tx_pkt(struct net_buf *nb)
+	cpd = conn_param_data_get(conn);
+	ud = net_buf_user_data(nb);
+
+	if (cpd->id == 0 || cpd->id != ud->id) {
```
A command can be sent over 3 packets, if the device disconnects after receiving the first packet then cpd->id will be updated to 0, 0 meaning invalid, and this check will then hit. 0 is reserved for "invalid ID", it specifically means that the connection is not valid.

nordicjm commented:
Reverted back to an incrementing counter as I don't see the value in tick count, it adds nothing useful and still needs the same != 0 check. Have allowed cpd to be null instead of asserting in the get function and added null checks and LOG_ERR() output on it being null, moved ID generation to alloc function.

jori-nordic reviewed (approved):
Advertising is getting some attention recently in the host, so I'll make a task to fix this behavior properly (ie restart advertising after the only conn obj is freed).

carlescufi reviewed (changes requested):
Please add an entry in the release notes

nordicjm commented:
Release notes on https://github.com/zephyrproject-rtos/zephyr/pull/52098

de-nordic reviewed (approved):
The new callbacks look OK.
I do not like that we have to generate own unique IDs for connections, as it should be done by BT, but it is only workaround we can afford now.

cfriedt commented:
@nordicjm - in order to backport this change, the `backport v2.7-branch` label should be added to this PR. Is there a reason that wasn't done? I should mention, that even if an automated backport fails, you can still modify it (it gets a special backport branch).

nordicjm commented:
@cfriedt I didn't do it from this because the code in 2.7 is so vastly outdated compared to main (e.g. different filenames and paths, different functions, different code) that none of it applies and it was just easier to manually copy and paste the bits into the 2.7 tree
